File: python/structural/decorator/decorators/encrypt_message.py

from decorators.message_decorator import MessageDecorator
from message.message import IMessage

from Crypto.Cipher import AES
from Crypto.Hash import SHA256
from Crypto.Protocol.KDF import PBKDF2
from Crypto.Random import get_random_bytes

import base64

class EncryptMessage(MessageDecorator):
    def __init__(self, user: str, password: str, message: IMessage):
        self.user = user
        self.password = password
        super().__init__(message)

    def encrypt_message(self) -> 'IMessage':
        try:
            #1. Obtener el contenido
            plaintext = self.message.get_content().encode('utf-8')

            #2. Derivar una clave segura (256 bits)
            salt = get_random_bytes(16)
            key = PBKDF2(self.password, salt, dkLen=32, count=100_000, hmac_hash_module=SHA256)

            #3. Crear el cifrado con AES en modo EAX
            cipher = AES.new(key, AES.MODE_EAX)
            nonce = cipher.nonce
            ciphertext, tag = cipher.encrypt_and_digest(plaintext)


            #4. Empaquetar todo (salt + nonce + tag + ciphertext) y pasarlo a base64
            encrypt_data = salt + nonce + tag + ciphertext
            encrypt_b64 = base64.b64encode(encrypt_data).decode('utf-8')

            #5. Guardar el contenido en un decorador
            self.set_content(encrypt_b64)

            return self.message

        except Exception as error:
            raise RuntimeError("Encrypted failed.") from error
        
    def process_message(self) -> 'IMessage':
        self.message = self.message.process_message()
        return self.encrypt_message()
    

        


    

                

File: typescript/src/structural/decorator/decorators/EncryptMessage.ts
import { IMessage } from "../interfaces/IMessage";
import { MessageDecorator } from "./MessageDecorator";

import { Buffer } from "buffer";
import crypto from "crypto"

export class EncryptMessage extends MessageDecorator {
    private password: string;

    constructor( password: string, message: IMessage){
        super(message)
        this.password = password
    }

    encryptMessage(): IMessage{
        try{

            const plaintext = Buffer.from(this.message.getContent(), "utf-8");
            
            const salt = crypto.randomBytes(12)
            const key = crypto.pbkdf2Sync(this.password, salt, 100000, 32 , "sha256")
            
            const cipher = crypto.createCipheriv("aes-256-ccm", key, salt, {authTagLength: 16});
            const ciphertext = Buffer.concat([cipher.update(plaintext), cipher.final()]);
            const tag = cipher.getAuthTag();

            const encryptData = Buffer.concat([salt, tag, ciphertext]);
            const encryptB64 = encryptData.toString("base64");

            this.message.setContent(encryptB64);

            return this.message;
        }
        catch(error){
            throw new Error("Encryption failed: "+ error)
        }

    }

    processMessage(): IMessage {
        this.message = this.message.processMessage();
        return this.encryptMessage();
    }

}